plugin/storage/factory.go

```diff
@@ -290,45 +330,57 @@ func (f *Factory) initDownsamplingFromViper(v *viper.Viper) {
 	f.FactoryConfig.DownsamplingHashSalt = v.GetString(downsamplingHashSalt)
 }
 
-// CreateArchiveSpanReader implements storage.ArchiveFactory
-func (f *Factory) CreateArchiveSpanReader() (spanstore.Reader, error) {
-	factory, ok := f.factories[f.SpanReaderType]
+func (f *Factory) createArchiveSpanReader() (spanstore.Reader, error) {
+	factory, ok := f.archiveFactories[f.SpanReaderType]
 	if !ok {
 		return nil, fmt.Errorf("no %s backend registered for span store", f.SpanReaderType)
 	}
-	archive, ok := factory.(storage.ArchiveFactory)
-	if !ok {
-		return nil, storage.ErrArchiveStorageNotSupported
-	}
-	return archive.CreateArchiveSpanReader()
+	return factory.CreateSpanReader()
 }
 
-// CreateArchiveSpanWriter implements storage.ArchiveFactory
-func (f *Factory) CreateArchiveSpanWriter() (spanstore.Writer, error) {
-	factory, ok := f.factories[f.SpanWriterTypes[0]]
+func (f *Factory) createArchiveSpanWriter() (spanstore.Writer, error) {
+	factory, ok := f.archiveFactories[f.SpanWriterTypes[0]]
 	if !ok {
 		return nil, fmt.Errorf("no %s backend registered for span store", f.SpanWriterTypes[0])
 	}
-	archive, ok := factory.(storage.ArchiveFactory)
-	if !ok {
-		return nil, storage.ErrArchiveStorageNotSupported
+	return factory.CreateSpanWriter()
+}
+
+func (f *Factory) InitArchiveStorage(
+	logger *zap.Logger,
+) (spanstore.Reader, spanstore.Writer) {
+	reader, err := f.createArchiveSpanReader()
+	if err != nil {
+		logger.Error("Cannot init archive storage reader", zap.Error(err))
+		return nil, nil
 	}
-	return archive.CreateArchiveSpanWriter()
+	writer, err := f.createArchiveSpanWriter()
+	if err != nil {
+		logger.Error("Cannot init archive storage writer", zap.Error(err))
+		return nil, nil
+	}
+
+	return reader, writer
 }
 
 var _ io.Closer = (*Factory)(nil)
 
 // Close closes the resources held by the factory
 func (f *Factory) Close() error {
 	var errs []error
+	closeFactory := func(factory storage.Factory) {
+		if closer, ok := factory.(io.Closer); ok {
+			if err := closer.Close(); err != nil {
+				errs = append(errs, err)
+			}
+		}
+	}
 	for _, storageType := range f.SpanWriterTypes {
```

Comment:
Any reason why this shouldn't simply be looping through `f.factories` and `f.archiveFactories` and closing them all? Seems weird that it depends on external list of types, considering that some types are added implicitly. Unless the same storage can be added more than once.

Reply:
@yurishkuro Yeah I think that would be more correct - should we do this in a follow-up PR?